### ISSUE TITLE
Configure OpenMPI with Torque

### ISSUE THREAD

**Issue by ealemany** [open-mpi/ompi#8085]
## Background information
Thank you for looking into this issue. I have a small cluster with a headnode and two compute nodes. All servers are running CentOS 7. I have been asked to install and configure Open MPI with Torque. Torque works fine we are able to submit jobs to the compute nodes.

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
Open MPI version is 4.0.5
 
### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

I have installed Open MPI from source.

### Please describe the system on which you are running

* Operating system/version: CentOS 7
* Computer hardware: Headnode Intel - Compute nodes AMD
* Network type: TCP - NFS 

-----------------------------

## Details of the problem

I have install and configure Open MPI with the following options "--prefix=/opt/openmpi-4.0.5 --with-tm --enable-orterun-prefix-by-default". after the installation i ran "ompi_info |  grep tm" on all the servers to make that "tm" was installed.
```
$ ompi_info |  grep tm
  Configure command line: '--prefix=/opt/openmpi-4.0.5' '--with-tm' '--enable-orterun-prefix-by-default'
                 MCA ess: tm (MCA v2.1.0, API v3.0.0, Component v4.0.5)
                 MCA plm: tm (MCA v2.1.0, API v2.0.0, Component v4.0.5)
                 MCA ras: tm (MCA v2.1.0, API v2.0.0, Component v4.0.5)
```

I can run simple "mpirun" test like the ones in the example of Open MPI directory.
```
$  mpirun --hostfile /opt/hostfile -np 64 --map-by node hostname
```
I even ran a "hello_c" program with an infinite loop in the code so i can "htop" on all the servers and see the allocated CPU's.

I also ran the ring_c program which sends a simple MPI message around a ring 10 times.
```
$ mpirun  --hostfile /opt/hostfile -np 64 /opt/openmpi-4.0.5/examples/ring
Warning: Permanently added 'tq3,10.112.0.14' (ECDSA) to the list of known hosts.
Warning: Permanently added 'tq4,10.112.0.16' (ECDSA) to the list of known hosts.
Process 0 sending 10 to 1, tag 201 (64 processes in ring)
Process 0 sent to 1
Process 0 decremented value: 9
Process 0 decremented value: 8
Process 0 decremented value: 7
Process 0 decremented value: 6
Process 0 decremented value: 5
Process 0 decremented value: 4
Process 0 decremented value: 3
Process 0 decremented value: 2
Process 0 decremented value: 1
Process 0 decremented value: 0
Process 0 exiting
Process 1 exiting
etc..............
```

However, The issue i am having is when I submit "mpirun" job to Torque only 1 core per server is utilized. Although, I specifically asked for all the cores in the cluster (64 cores - 32 cores per compute node).

Here is my shell script:
```
!/bin/bash
#PBS -k o
#PBS -l nodes=2:ppn=32,walltime=12:00:00
#PBS -W x=nmatchpolicy:exactnode
#PBS -N mpi_test_tq3_4

mpirun   /opt/openmpi-4.0.5/examples/ring
```

This is how i submit the job:
```
$ qsub -l nodes=2:ppn=32 -W x=nmatchpolicy:exactnode mpi_test.sh
49.mt-manager
```
The output file shows that only 2 processors were allocated.
```
$ cat  mpi_test_tq3_4.o49
Process 0 sending 10 to 1, tag 201 (2 processes in ring)
Process 0 sent to 1
Process 0 decremented value: 9
Process 0 decremented value: 8
Process 0 decremented value: 7
Process 0 decremented value: 6
Process 0 decremented value: 5
Process 0 decremented value: 4
Process 0 decremented value: 3
Process 0 decremented value: 2
Process 0 decremented value: 1
Process 0 decremented value: 0
Process 0 exiting
Process 1 exiting
```
We can see that only 1 processor was allocated to the job "Process 0 sending 10 to 1, tag 201 (2 processes in ring)"
Did I do something wrong when I configured Open MPI with the "--with-tm" ?
Should I have include the path to the PBS Torque install directory (--wiht-tm=/path/to/torque/directory) ?
Am i submitting the job with the proper $PBS option and "qsub" options ?

Thank you in advance for your help in this issue.

Eric

**Comment by cniethammer** [comment]
Open MPI will try to figure out the location of the Torque libraries and headers using `pbs-config`. If this is not present (e.g. as in PBS Professional) in your installation you will have to specify the Torque installation path.

**Comment by ggouaillardet** [comment]
Since `--with-tm` is (explicitly) used in the `configure` command line, the expected behavior is to fail if the Torque/PBSPro libraries/headers cannot be found.

**Comment by jsquyres** [comment]
According to @ealemany's output, Open MPI did build and install with TM support.

@ealemany: you might want to try a few things to ensure that a) the TM support is being selected, and b) the TM plugins are loading properly at run time.

For example, try submitting a Torque job with:

```
mpirun --mca plm tm ...
```

This will tell Open MPI to _only_ use the `tm` PLM plugin.  Specifically: it will fail if the `tm` PLM plugin cannot be used.

You may also want to look at the output of

```
env | sort
```

in your Torque job, and ensure that you see a whole pile of PBS/Torque environment variables.  The `tm` PLM isn't using those (it's actually using the Torque TM API), but seeing them there would be a good indication that information is flowing through your PBS/Torque environment properly in jobs that are launched.

**Comment by ealemany** [comment]
@jsquyres - Thank you.

What do you mean when you say 
"you may also want to look at the output of 
```
env | sort
```
in your Torque job".  
I submitted a Torque job with 
```
mpirun --mca plm  /path/to/my/ring-program
```
which failed but i do not know how to "env | sort" in the Torque job.

Thank you

**Comment by jsquyres** [comment]
@ealemany What is the output from when `mpirun --mca plm /path/to/ring` failed?

In your script, just put:

```
!/bin/bash
#PBS -k o
#PBS -l nodes=2:ppn=32,walltime=12:00:00
#PBS -W x=nmatchpolicy:exactnode
#PBS -N mpi_test_tq3_4

echo "=== This is the environment"
env | sort
echo "=== End of environment"

mpirun   /opt/openmpi-4.0.5/examples/ring
```

**Comment by ealemany** [comment]
Thank you for the example @jsquyres 
So, I think i have a bigger problem than what i think and may not be related to Open MPI.
When i run a job allocating 2 nodes and 8 cores per nodes - the job runs fine.
```
#!/bin/bash
#PBS -k o
#PBS -l nodes=2:ppn=8 -W x=nmatchpolicy:exactnode
#PBS -m abe
#PBS -M ealemany@stanford.edu
#PBS -N mpi_test_tq3_4
#PBS -j oe


echo "=== This is the environment"
env | sort
echo "=== End of environment"

mpirun /opt/openmpi-4.0.5/examples/ring
```
output
```
=== This is the environment
_=/bin/env
ENVIRONMENT=BATCH
HISTCONTROL=ignoredups
HISTSIZE=1000
HISTTIMEFORMAT=%d/%m/%y %T
HOME=/home/eric
HOSTNAME=tq3.stanford.edu
LANG=en_US.UTF-8
LD_LIBRARY_PATH=:/opt/openmpi-4.0.5/lib
LESSOPEN=||/usr/bin/lesspipe.sh %s
LOGNAME=eric
MAIL=/var/spool/mail/eric
PATH=/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/openmpi-4.0.5/bin:/home/eric/.local/bin:/home/eric/bin
PBS_ENVIRONMENT=PBS_BATCH
PBS_GPUFILE=/var/spool/torque/aux//112.mt-manager.stanford.edugpu
PBS_JOBCOOKIE=4D43C4F7A261B6563147F1C9AF82E339
PBS_JOBID=112.mt-manager.stanford.edu
PBS_JOBNAME=mpi_test_tq3_4
PBS_MICFILE=/var/spool/torque/aux//112.mt-manager.stanford.edumic
PBS_MOMPORT=15003
PBS_NODEFILE=/var/spool/torque/aux//112.mt-manager.stanford.edu
PBS_NODENUM=0
PBS_NP=16
PBS_NUM_NODES=2
PBS_NUM_PPN=8
PBS_O_HOME=/home/eric
PBS_O_HOST=mt-manager.stanford.edu
PBS_O_LANG=en_US.UTF-8
PBS_O_LOGNAME=eric
PBS_O_MAIL=/var/spool/mail/root
PBS_O_PATH=/usr/local/bin:/usr/local/sbin:/opt/moab/bin:/usr/local/bin:/usr/local/sbin:/opt/moab/sbin:/opt/moab/bin:“/opt/openmpi-4.0.5/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games":/opt/moab/bin:/opt/moab/sbin:/opt/openmpi-4.0.5/bin:/root/bin:/opt/openmpi-4.0.5/bin
PBS_O_QUEUE=batch
PBS_O_SERVER=mt-manager.stanford.edu
PBS_O_SHELL=/bin/bash
PBS_O_WORKDIR=/home/eric
PBS_QUEUE=batch
PBS_TASKNUM=1
PBS_VERSION=TORQUE-6.1.3
PBS_VNODENUM=0
PBS_WALLTIME=3600
PWD=/home/eric
SHELL=/bin/bash
SHLVL=2
USER=eric
=== End of environment
Process 0 sending 10 to 1, tag 201 (16 processes in ring)
Process 0 sent to 1
Process 0 decremented value: 9
Process 0 decremented value: 8
Process 0 decremented value: 7
Process 0 decremented value: 6
Process 0 decremented value: 5
Process 0 decremented value: 4
Process 0 decremented value: 3
Process 0 decremented value: 2
Process 0 decremented value: 1
Process 0 decremented value: 0
Process 0 exiting
Process 2 exiting
Process 1 exiting
Process 4 exiting
Process 5 exiting
Process 6 exiting
Process 7 exiting
Process 3 exiting
Process 8 exiting
Process 9 exiting
Process 10 exiting
Process 11 exiting
Process 12 exiting
Process 13 exiting
Process 14 exiting
Process 15 exiting
```
But when I want to run a job allocating 2 nodes and 32 cores per nodes (which each node has 32cores) the job just keep restarting every second and never actually starts. I have to kill the job and there are no output files.
The "mpirun" command with the --hostfile option  runs the "ring" just fine when ran directly at the prompt.
```
[eric@mt-manager ~]$ mpirun  --hostfile /opt/hostfile /opt/openmpi-4.0.5/examples/ring
Warning: Permanently added 'tq3,10.112.0.14' (ECDSA) to the list of known hosts.
Warning: Permanently added 'tq4,10.112.0.16' (ECDSA) to the list of known hosts.
Process 0 sending 10 to 1, tag 201 (64 processes in ring)
Process 0 sent to 1
Process 0 decremented value: 9
Process 0 decremented value: 8
Process 0 decremented value: 7
Process 0 decremented value: 6
Process 0 decremented value: 5
Process 0 decremented value: 4
Process 0 decremented value: 3
Process 0 decremented value: 2
Process 0 decremented value: 1
Process 0 decremented value: 0
Process 0 exiting
Process 1 exiting
Process 2 exiting
Process 3 exiting
Process 4 exiting
Process 5 exiting
Process 6 exiting
Process 7 exiting
Process 32 exiting
Process 8 exiting
Process 33 exiting
Process 9 exiting
Process 34 exiting
Process 10 exiting
Process 35 exiting
"
"
===== all the way down to Process 63 exiting========
```
I think it is a Torque issue not being able to initiate the right amount of CPU (?)

**Comment by jsquyres** [comment]
Your output all looks good (you can ditch the `env ...` now -- the point of that was just to verify that the `PBS_*` env variables were there, and they are -- good!).

> But when I want to run a job allocating 2 nodes and 32 cores per nodes (which each node has 32cores) the job just keep restarting every second and never actually starts. I have to kill the job and there are no output files.

That is, indeed, odd.

You might try mpirun'ing `hostname` (i.e., just the Linux `hostname` command -- it's not an MPI program at all).  See if that is able to run in a 2 x 32 Torque job.  Torque + Open MPI should be capable of running at significantly larger scale than that.

It would be good to know _why_ your Torque job keeps restarting.  Is it failing, and therefore the Torque manager re-queues the job?  If so, why/how?

**Comment by ealemany** [comment]
@jsquyres I tried running "mpirun hostname" in a Torque job as you suggested with 2 nodes 32 cores each and got the same result as before. When i run "qstat" the job is in "R" mode but no timing. Then the job goes into "Q" mode for a second and back to "R" mode and so on... The job does not "fail" per say since it never starts. I really think it is a CPU configuration at lower level of the system. The reason why i am saying that is because on the 32core nodes when i looked at the "cpuset.mems" in "/sys/fs/cgroup/cpuset" it shows "0,2" instead it should say "0-3"
```
[root@tq4 cpuset]# cat cpuset.mems
0,2
```
Also, when i reboot the system the I can see that Node0 and Node2 are enabled but Node1 and Node3 are not. I think that's why i can only run a job with 16 cores (8 cores per channel).
![IMG_2559](https://user-images.githubusercontent.com/36861498/96297277-7b71ad00-0fa5-11eb-928e-2794d285b6f3.JPG)
 
Somehow I need to find a way to enable Node1 and Node3. I don't see anything in the BIOS that allows to enable Node1 and Node3. I updated the BIOS and still no change. 

I think this discussion is going beyond this thread. I will probably need to reach out to a AMD CPU "guru".
